File: src/lowlevel/client/icefsoperators/icefsAccess.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:39
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsAccess.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsAccess(fuse_req_t fuseReq, fuse_ino_t inode,
                                int mask) {
  ICEFS_PR_FUNCTION;
  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsAccessReq req;
      icefsgrpc::IcefsAccessRes res;
      grpc::ClientContext ctx;

      req.set_inode(inode);
      req.set_mask(mask);

      grpc::Status status = gRpcClient->DoIcefsAccess(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsAccessReq req;
      icefsthrift::IcefsAccessRes res;

      req.__set_inode(inode);
      req.__set_mask(mask);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsAccess(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsAccess.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-17 14:22:56
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsAccess.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"fmt"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsAccess(fakeInode uint64, mask int32) (status int32) {
	var procName string
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		return icefserror.ICEFS_BUG_ERR
	}

	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	procName = fmt.Sprintf("/proc/self/fd/%v", inode.fd)
	status = icefserror.IcefsStdErrno(syscall.Access(procName, uint32(mask)))
	inode.inodeLock.RUnlock()
	return
}

func (s *IcefsGRpcServer) DoIcefsAccess(ctx context.Context, req *pb.IcefsAccessReq) (*pb.IcefsAccessRes, error) {
	var res pb.IcefsAccessRes

	res.Status = s.server.doIcefsAccess(req.Inode, req.Mask)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsAccess(ctx context.Context, req *icefsthrift.IcefsAccessReq) (*icefsthrift.IcefsAccessRes, error) {
	var res icefsthrift.IcefsAccessRes

	res.Status = s.server.doIcefsAccess(uint64(req.Inode), req.Mask)

	return &res, nil
}
